src/main/java/io/jenkins/dockerjavaapi/client/DelegatingDockerClient.java

```diff
@@ -0,0 +1,465 @@
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+
+package io.jenkins.dockerjavaapi.client;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import javax.annotation.Nonnull;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.command.*;
+import com.github.dockerjava.api.exception.DockerException;
+import com.github.dockerjava.api.model.*;
+
+// MAINTENANCE NOTE:
+// The DockerClient API varies depending on the version of the docker-java
+// library that this plugin provides.
+// It may be necessary to add/remove methods from this class when changing
+// the version of docker-java.
+// ...but, by having this here, it allows all other Jenkins plugins to be much
+// less affected by changes to the version of docker-java, meaning that this
+// plugin can be upgraded without breaking every other docker-using plugin.
+
+/**
+ * Simple delegate class for the {@link DockerClient} interface.
+ * <p>
+ * This makes it easy for other classes to override specific methods without
+ * having to implement all of them. Every method here:
+ * <ul>
+ * <li>calls {@link #getDelegate()},</li>
+ * <li>calls the matching method on the delegate class,</li>
+ * <li>for methods that return void, calls {@link #interceptVoid()} and then
+ * returns,</li>
+ * <li>for other methods, calls {@link #interceptAnswer(Object)}, passing in the
+ * delegate method's answer and returns whatever
+ * {@link #interceptAnswer(Object)} returned.</li>
+ * </ul>
+ * </p>
+ * If you are writing a Jenkins plugin that needs a class to implement/wrap
+ * {@link DockerClient}, you'd be best advised to extend this one, otherwise
+ * your code could fail whenever the version of this plugin changes.
+ */
+@SuppressWarnings("deprecation")
+public class DelegatingDockerClient implements DockerClient {
+
+    private final DockerClient delegate;
+
+    /**
+     * Constructs a new instance that delegates all API calls to the specified
+     * {@link DockerClient}.
+     *
+     * @param delegate The {@link DockerClient} to delegate to.
+     */
+    public DelegatingDockerClient(@Nonnull DockerClient delegate) {
+        this.delegate = delegate;
+    }
+
+    /**
+     * Obtains the underlying {@link DockerClient} interface. Subclasses can
+     * override this if they need to hook into every call.
+     *
+     * @return the {@link DockerClient} to be delegated to.
+     */
+    @Nonnull
+    protected DockerClient getDelegate() {
+        return delegate;
+    }
+
+    /**
+     * Called just before the result is returned. Allows a subclass to act just
+     * before the method returns and/or to alter the result.
+     * 
+     * @param originalAnswer The result from the delegate.
+     * @return The result to be returned instead.
+     */
+    protected <T> T interceptAnswer(T originalAnswer) {
+        return originalAnswer;

```

Comment:
While I perfectly understand the interest of this class, I do not get if for this method ...

I might be wrong, but since it is generic, one will not be able to overload it with different signatures :
```
protected AttachContainerCmd interceptAnswer(AttachContainerCmd original) { ... }
protected AuthCmd interceptAnswer(AuthCmd original) { ... }
```
The only option is to use an ugly switch / case based on parameter class.

While you could just overload the public method you need to :
```
@Override
public AuthCmd authCmd() {
  AuthCmd result = super.authCmd();
  // do whatever you will have done in interceptAnswer()
  return result;
}
```

Reply:
I agree that it will be difficult to do anything "method specific" at that point, and so if you're doing anything "method specific" you're better off overriding the method ... but my rationale here was that the `interceptAnswer` method is here to allow a plugin to intercept methods that _weren't part of the API when they were coded._

e.g. a plugin could implement `interceptAnswer` to throw some sort of NotPermittedException on all methods but override the few methods they wanted to permit to let those operate normally.
e.g. a plugin that needed to only tinker with container creation could use `if (originalAnswer instanceof CreateContainerCmd)` clauses and have their code continue to work even if new `createContainerCmd` methods were added to the API at a later date.
i.e. it provides a way of coping with the API changing without forcing everyone to re-code (the plugins outside of this one) every time this plugin changes.

TL;DR: The "interceptAnswer" method is like using a "default" clause on a switch(someEnum)

Reply:
ok, understood.

However, I still do not see clearly how all of this will work, having the API updates but not the plugin, or the opposite, ...
I fear binary compatibility problems will still occur.

Let's see !

Reply:
This plugin is 100% in control over what version of the `docker-java` library is on the classpath in Jenkins.
The problem I (personally) have is that the docker-plugin needs to be able to use a class like `DelegatingDockerClient` (and currently has its own implementation) but it (the docker-plugin) does not control the version of the `docker-java` library, so whenever _this_ plugin changes to update the docker-java library, the docker-plugin stops working (and gets compilation errors) until it's changed to match.
i.e. we have a tight dependency between two supposed-to-be-independent plugins - they either both change, or neither.

My _theory_ is that by having `DelegatingDockerClient` code in *this* plugin (that's tightly bound to the `DockerClient` API from the docker-java library, whose version is also decided by *this* plugin), another plugin (like the docker-plugin) can "simply" use this plugin's `DelegatingDockerClient` safe in the knowledge that, even if the `docker-java` library is later changed to a different version, this plugin will have made matching changes to `DelegatingDockerClient`, so everything (including the docker-plugin) should keep working (and thus reducing stress & workload on plugin maintainers).

i.e. we're co-locating "control of the DockerClient API" with "code that needs to change every time the DockerClient API changes" so that, hopefully, nobody else needs to worry about it again 🤞 

PS. This should keep working until the `DockerClient` changes its API sufficiently to be entirely incompatible, but that seems to happen less often than it merely being extended.  If we do encounter that issue often enough that it becomes a big problem then we could put in additional code here to bridge the gap between the docker-java versions (by making this plugin responsible for presenting a consistent API to Jenkins), but I figure we can cross that bridge when we come to it...

Reply:
agreed and convinced ;)